---
layout: lesson
root: .  # Is the only page that doesn't follow the pattern /:path/index.html
permalink: index.html  # Is the only page that doesn't follow the pattern /:path/index.html
---

> ## Prerequisites
>
> All pre-requisites are taken care of if you are using this lesson at the Otago Bioinformatics Spring School 2021.
>
> Please see the [setup](setup.md) instructions if you wish to undertake this workshop on your own computer.
{: .prereq}

## About

Advances in genomics over the past decade have greatly expanded the breadth of community ecology, and enabled fine scale study of many ecosystems. Environmental DNA (eDNA) is DNA material isolated from environmental substrates, including water, soil, air, or even within larger organisms (using a broader definition of ‘environment’). The eDNA samples can be amplified with PCR and sequenced with next-generation sequencing technologies to study the biodiversity of the environment, in a method called metabarcoding. In this workshop we will review the most common methods for metabarcoding in order to identify the taxonomic components and determine the relationship across these communities. Through a combination of demonstration and hands-on practicals, you will learn the basic concepts and core tools needed to analyse eDNA metabarcoding data for a wide range of studies.

The material for this course was developed by Hugh Cross and Gert-Jan Jeunen. Valuable contributions and suggestions were made by Ngoni Faya, Lucie Jiraska, Dinindu Senanayake, and Annabel Whibley, all of whom also helped teach the material.

This workshop is sponsored by <a href="https://www.genomics-aotearoa.org.nz/" target="_blank" rel="noopener noreferrer">Genomics Aotearoa</a> and the <a href="https://www.nesi.org.nz/" target="_blank" rel="noopener noreferrer">New Zealand eScience Infrastructure (NeSI)</a>. 




{% include links.md %}
